QUARKUS-2812.md

```diff
@@ -0,0 +1,51 @@
+# QUARKUS-2812 Build time metrics for Quarkus
+
+JIRA link: https://issues.redhat.com/browse/QUARKUS-2812.
+
+The aim of this feature is to collect build-time data in a Quarkus application and send them to an external data platform
+for analytical purposes.
+
+## Scope of testing
+
+What data is collected in what event.
+- Events:
+  - Dev mode build - should send a minimal data payload (a "ping").
+  - Production build (JVM, native) - should send full analytics data.
+- Data:
+  - Build type.
+  - Quarkus version, Java vendor and version, GraalVM/Mandrel version, OS name and version, build tool and its version (Gradle, Maven).
+  - Anonymous User ID, anonymous Project ID.
+  - List of extensions and their versions.
+  - Build duration, build errors
+
+Verify the correct destination of the data.
+- How to test that the correct data has been collected. Mock the destination, query it?
+
+Verify, that no sensitive data / personal identifiable information is being shared (e.g. IP address).
+
+Verify that data cannot be sent and read without proper authentication.
+
+Verify that there is a working way of turning the telemetry off.
+- A system property.
+
+Verify that a clear notice about the telemetry and the nature of the collected data is part of Quarkus release.
+- Documentation.
+- Log message about the telemetry being active/inactive.
+
+## Automated test development
+
+`StartStopTS` is a suitable place for adding the tests.
+- Various application setups.
+  - Sets of extensions.
+  - Build-time analytics feature on/off.
+- Various build types.
+
+Implement a (REST) client which will be able to query the external data platform and verify the stored data for a given User ID / Project ID.
```

Comment:
This was one of the discussion topics on Monday. We should be able to get the dump of the payload sent to the service. Users and customers can want to audit the data that are sent to the service.

Depending on the real service can be a tricky too. There can be outages, the service can be overloaded and the payload can be put into a queue for processing a bit later etc.

Reply:
Planned automated tests changed to work with the data payload log, instead of the remote service.